### PR TITLE
PRBS RX State Data Type

### PR DESCRIPTION
**Pull request by JaiOCP**
Read Only attribute for getting Rx error count

### CODE REVIEW

**Review by kcudnik — APPROVED** [review]

File: inc/saiport.h [hunk 1]
```diff
@@ -1501,6 +1501,14 @@ typedef enum _sai_port_attr_t
      */
     SAI_PORT_ATTR_PRBS_POLYNOMIAL,
 
+    /**
+     * @brief Port PRBS receive error count
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_PRBS_RX_ERROR_COUNT,
```
Comment (lguohan) [inline inc/saiport.h:1510]: can this be in the port stat, rather than port attribute?

**Comment by rlhui** [comment]
@marian-pritsak - would you like to review this for MLNX?

File: inc/saiport.h (hunk 1 above)
Comment (JaiOCP) [reply inc/saiport.h]: move to port stats
https://github.com/opencomputeproject/SAI/pull/1179/commits/bed90dd32bff2db5c88a0821a7ac2f050c199b79

File: inc/saiport.h [hunk 2]
```diff
@@ -2290,6 +2290,9 @@ typedef enum _sai_port_stat_t
     /** PRBS Error Count */
     SAI_PORT_STAT_PRBS_ERROR_COUNT,
 
+    /** PRBS Receive Error Count */
+    SAI_PORT_STAT_PRBS_RX_ERROR_COUNT,
```
Comment (lguohan) [inline inc/saiport.h:2294]: what is the existing SAI_PORT_STAT_PRBS_ERROR_COUNT?

Comment (JaiOCP) [reply inc/saiport.h]: Removed this attribute
https://github.com/opencomputeproject/SAI/pull/1179/commits/ce4f8429862738e2707537cd3459a30a08912321

Also introduced the complex data type for handling atomic state as discussed in the meeting

File: inc/saitypes.h [hunk 3]
```diff
@@ -391,6 +391,64 @@ typedef struct _sai_ip_prefix_t
...
+{
+    /** @validonly type == SAI_FLEX_DATA_TYPE_PRBS */
+    sai_port_prbs_rx_status_t rx_status;
+
+} sai_flex_status_t;
+
+/**
+ * @extraparam sai_flex_data_type_t type
+ */
+typedef union _sai_flex_data_t
```
Comment (kcudnik) [inline inc/saitypes.h:434]: why this is union ? and contains only 1 element ? are you planning to extend that ?

Comment (JaiOCP) [reply inc/saitypes.h:434]: Yes, thats the idea. This is a generic data type for atomic I/O and more fields will be added later as needed

**Review by kcudnik — APPROVED** [review]

**Comment by rlhui** [comment]
@rck-innovium please help review/approve, thanks.

File: inc/saitypes.h [hunk 4]
```diff
@@ -391,6 +391,64 @@ typedef struct _sai_ip_prefix_t
...
+{
+    /** @validonly type == SAI_FLEX_DATA_TYPE_PRBS */
+    uint32_t count;
+
+} sai_flex_data_t;
```
Comment (lguohan) [inline inc/saitypes.h:439]: we can define a specific attribute value type to combine prbs status and prbs error count in the new attribute value type.

Comment (JaiOCP) [reply inc/saitypes.h]: https://github.com/opencomputeproject/SAI/pull/1179/commits/65aec0b9d7be45f449561d25f1a8277bfff14bd3

Changed as per review comments

File: inc/saiport.h [hunk 5]
```diff
@@ -1566,6 +1547,14 @@ typedef enum _sai_port_attr_t
      */
     SAI_PORT_ATTR_PRBS_RX_STATUS,
 
+    /**
+     * @brief Attribute data for #SAI_PORT_ATTR_PRBS_RX_STATE
```
Comment (lguohan) [inline inc/saiport.h:1551]: add some comments here to explain why we introduce this attribute and not reuse existing attribute.

a simplified version of following para is good.

There are many “clear on read” states in hardware. Sometimes these are individual standalone fields. "Many a times HW combines multiple fields in a single atomic read. 
Given that SAI API attributes are defined at sub field granularity, SAI adapter has to emulate “clear on read” for sub field queries. Flex state complex data type allows read operation on the combination of such attributes. This makes SAI adapter implementation easy by just making a single query to hardware. Hardware natively takes care of clear on read for all the subfields with in the field.

Comment (JaiOCP) [reply inc/saiport.h:1551]: https://github.com/opencomputeproject/SAI/pull/1179/commits/97ed2bb8e433f7ce80e9e898c6c75c4d8eff9d06

Done

**Review by lguohan — APPROVED** [review]

**Review by mikeberesford — APPROVED** [review]

**Comment by JaiOCP** [comment]
@rlhui Please merge

File: inc/saiport.h [hunk 6]
```diff
@@ -1566,6 +1547,16 @@ typedef enum _sai_port_attr_t
...
+     * Used for clear on read status/count register.
+     * Adapter should return return SAI_STATUS_NOT_SUPPORTED if not supported.
```
Comment (rck-innovium) [inline inc/saiport.h:1553]: typo: return return

**Review by rck-innovium — CHANGES_REQUESTED** [review]
Hi Jai,

We need the below understanding to be documented:

----
There are multiple SAI attributes/stat-id (viz., SAI_PORT_ATTR_PRBS_RX_STATE,  SAI_PORT_ATTR_PRBS_RX_STATUS, SAI_PORT_PRBS_RX_STATUS_LOCK_WITH_ERRORS) for PORT PRBS status. It depends on the underlying ASIC as to which of these attributes are applicable. The suggestion was that the application would first query  SAI_PORT_ATTR_PRBS_RX_STATE. If the SAI implementation returns that the attribute is unsupported, the application would then fallback to query the existing SAI_PORT_ATTR_PRBS_RX_STATUS and SAI_PORT_PRBS_RX_STATUS_LOCK_WITH_ERRORS.

----

File: inc/saiport.h (hunk 6 above)
Comment (JaiOCP) [reply inc/saiport.h:1553]: > Hi Jai,
> 
> We need the below understanding to be documented:
> 
> There are multiple SAI attributes/stat-id (viz., SAI_PORT_ATTR_PRBS_RX_STATE, SAI_PORT_ATTR_PRBS_RX_STATUS, SAI_PORT_PRBS_RX_STATUS_LOCK_WITH_ERRORS) for PORT PRBS status. It depends on the underlying ASIC as to which of these attributes are applicable. The suggestion was that the application would first query SAI_PORT_ATTR_PRBS_RX_STATE. If the SAI implementation returns that the attribute is unsupported, the application would then fallback to query the existing SAI_PORT_ATTR_PRBS_RX_STATUS and SAI_PORT_PRBS_RX_STATUS_LOCK_WITH_ERRORS.

Comments in header file are limited in characters.
We will add this detail in release notes

Comment (JaiOCP) [reply inc/saiport.h]: > typo: return return
Fixed
https://github.com/opencomputeproject/SAI/pull/1179/commits/e276ffafbc70c59d7171dfe4ce5e22b90d25ddfa

**Comment by rlhui** [comment]
> Hi Jai,
> 
> We need the below understanding to be documented:
> 
> There are multiple SAI attributes/stat-id (viz., SAI_PORT_ATTR_PRBS_RX_STATE, SAI_PORT_ATTR_PRBS_RX_STATUS, SAI_PORT_PRBS_RX_STATUS_LOCK_WITH_ERRORS) for PORT PRBS status. It depends on the underlying ASIC as to which of these attributes are applicable. The suggestion was that the application would first query SAI_PORT_ATTR_PRBS_RX_STATE. If the SAI implementation returns that the attribute is unsupported, the application would then fallback to query the existing SAI_PORT_ATTR_PRBS_RX_STATUS and SAI_PORT_PRBS_RX_STATUS_LOCK_WITH_ERRORS.

@rck-innovium  - do you agree/approve this PR?

**Comment by rck-innovium** [comment]
> @rck-innovium - do you agree/approve this PR?

Yes. Jai has said that he will document the NOS expectation in the release nots.

**Review by rck-innovium — APPROVED** [review]

**Pull request merged**